### ISSUE TITLE
Debug LabMetadata processing

### ISSUE THREAD

**Issue by victorskl** [umccr/data-portal-apis#223]
There seems to be issue with LabMetadata sheet processing. Library ID `L2100499` is, in fact, no duplicate in the Sheet. However, it raises `IntegrityError` as follows. Can check logs in respective scheduled lambda CloudWatch logs in both DEV and PROD.

Actions:
- Try to reproduce it locally (so far, can't reproduce it on local yet)
- Check bulk create behaviour `LabMetadata.objects.bulk_create()`
- Check dataframe processing, try to simplify it, if any

```
[ERROR] IntegrityError: (1062, "Duplicate entry 'L2100499' for key 'library_id'")
Traceback (most recent call last):
  File "/var/task/data_processors/lims/lambdas/labmetadata.py", line 56, in scheduled_update_handler
    return services.persist_labmetadata(df)
  File "/var/lang/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/var/task/data_processors/lims/services.py", line 179, in persist_labmetadata
    rows_created = LabMetadata.objects.bulk_create(instances_insert, batch_size=100)
  File "/opt/python/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/opt/python/django/db/models/query.py", line 514, in bulk_create
    returned_columns = self._batched_insert(
  File "/opt/python/django/db/models/query.py", line 1293, in _batched_insert
    self._insert(item, fields=fields, using=self.db, ignore_conflicts=ignore_conflicts)
  File "/opt/python/django/db/models/query.py", line 1270, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/opt/python/django/db/models/sql/compiler.py", line 1416, in execute_sql
    cursor.execute(sql, params)
  File "/opt/python/aws_xray_sdk/ext/django/db.py", line 24, in execute
    result = super(DjangoXRayTracedCursor, self).execute(query, *args, **kwargs)
  File "/opt/python/aws_xray_sdk/core/models/subsegment.py", line 54, in __call__
    return self.recorder.record_subsegment(
  File "/opt/python/aws_xray_sdk/core/recorder.py", line 435, in record_subsegment
    return_value = wrapped(*args, **kwargs)
  File "/opt/python/aws_xray_sdk/ext/dbapi2.py", line 47, in execute
    return self.__wrapped__.execute(query, *args, **kwargs)
  File "/opt/python/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/opt/python/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/opt/python/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/opt/python/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/opt/python/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/opt/python/django/db/backends/mysql/base.py", line 73, in execute
    return self.cursor.execute(query, args)
  File "/opt/python/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "/opt/python/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "/opt/python/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "/opt/python/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "/opt/python/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "/opt/python/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "/opt/python/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "/opt/python/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
```

**Comment by victorskl** [comment]
**Quick fix:**

_And for the sake of current active Run(s) -- 160, 161, 51 -- for T/N workflow_ 

Empty the metadata table:
```
truncate table data_portal_labmetadata;
```

And reprocess:
```
aws lambda invoke --profile prod \
  --function-name data-portal-api-prod-labmetadata_scheduled_update_processor \
  --cli-binary-format raw-in-base64-out \
  --payload '{"sheet": "2020"}' \
  prod_2020.json

aws lambda invoke --profile prod \
  --function-name data-portal-api-prod-labmetadata_scheduled_update_processor \
  --cli-binary-format raw-in-base64-out \
  --payload '{"sheet": "2021"}' \
  prod_2021.json
```

**Comment by victorskl** [comment]
This is not reproduce on subsequence daily runs, since. Keep watching it for now.

**Comment by victorskl** [comment]
Error happen again; last success processed was on 2021/06/24. 

CW log streams in PROD:
```
2021/06/24/[$LATEST]e68fcf59288946b68c5f605a88ac6006
2021/06/25/[$LATEST]963c98846be941d3b65186470f54afc5
2021/06/26/[$LATEST]2a26b576575e4ac0b96467d5b01c4dcd
2021/06/27/[$LATEST]7f5f05b526c24cf792decb895e866254
2021/06/28/[$LATEST]9231893c173243e58559057127bc1228
2021/06/29/[$LATEST]65e613fcba704fecba21a3372d8ff23e
2021/06/30/[$LATEST]a4f182fd7045435095c0c32c2ba3eb6c
```

As well as in DEV.

Action items:
- Apply quick fix ☝️ again
- I will revisit on this and improve processing, if any.

FYI @reisingerf

**Comment by reisingerf** [comment]
Happy to jump in too. Ping me when you start looking at it.

**Comment by victorskl** [comment]
Error recurring:

ENV: PROD
CW log group: `/aws/lambda/data-portal-api-prod-labmetadata_scheduled_update_processor`
CW log streams:
```
2021/08/26/[$LATEST]cf540cd23c0c44f0a1799f2c39fbe3b6
2021/08/27/[$LATEST]1b46120f0a2647d0a26abd37e976ce6c
2021/08/28/[$LATEST]8df04b2611f74a05a7501c1f589e20db
2021/08/29/[$LATEST]a874387a60204607a496bf384d217e10
```

Action items:

- Apply quick fix ☝️ and reproc logstream: `2021/08/30/[$LATEST]26f769ce89eb4d3183d479eaba16e942`

**Comment by victorskl** [comment]
Error recurring:
- LabMetadata import didn't work since last manual intervention ☝️ 

ENV: PROD
CW log group: `/aws/lambda/data-portal-api-prod-labmetadata_scheduled_update_processor`
CW log streams:
```
2021/08/30/[$LATEST]6414bd64af7b45208b55202e2fb99a06
2021/08/31/[$LATEST]dbbce26cf941427190be9a20de0e1f5c
2021/09/01/[$LATEST]f4b88dbf2f7146fcb833933586fd8f19
2021/09/02/[$LATEST]e866fdfc50e14fe8a8a7fabd56d6bf50
2021/09/03/[$LATEST]a0d301d09bc34169a01b42f185eeeb55
2021/09/04/[$LATEST]7ea7add9802a4a5eaa0ed6a6bf481103
```

Action items:

- Apply quick fix ☝️ 
- Reproc logstream: 
```
2021/09/05/[$LATEST]efc5792f6f3b4b2f97e1a08286e55125
```

**Effected and Blocked Run 170. Raised to Critical.**

FYI @reisingerf

**Comment by reisingerf** [comment]
TL;DR: change to replace instead of merge metadata

As this metadata is neither created nor changed in the portal, it's probably a fair strategy to leave the responsibility of completeness/consistency with the original data source (lab tracking sheet for now, Clarity in the near future). A simple replacement/refresh of the portal view from the external truth makes the import procedure much easier and cleaner and should fix the recent issues around metadata import/sync.

If doubts about external data integrity or consistency with portal data should arise, additional checks could be put in place at a later stage (and probably independent of the actual data loading) if/when needed.
